File: DSA/GRAPH/Basics/Representation/AdjacencyMatrixGraph.java
/*
 * ---------------- THEORY: Adjacency Matrix in Java ----------------
 *
 * An adjacency matrix is a 2D array used to represent a graph.
 * For a graph with V vertices:
 * - matrix[i][j] = 1     --> edge exists from vertex i to j (unweighted)
 * - matrix[i][j] = w     --> edge with weight w exists (weighted)
 * - matrix[i][j] = 0     --> no edge
 *
 * For undirected graphs:
 *   matrix[i][j] = matrix[j][i]
 *
 * Time Complexity:
 *   - Add edge: O(1)
 *   - Check edge: O(1)
 * Space Complexity: O(V^2)
 *
 * Suitable for dense graphs.
 */

import java.util.Scanner;

public class AdjacencyMatrixGraph {
    private int[][] matrix;
    private int vertices;
    private boolean isDirected;
    private boolean isWeighted;

    // Constructor
    public AdjacencyMatrixGraph(int vertices, boolean isDirected, boolean isWeighted) {
        this.vertices = vertices;
        this.isDirected = isDirected;
        this.isWeighted = isWeighted;
        matrix = new int[vertices][vertices]; // initialized to 0 by default
    }

    // Add edge between u and v (with optional weight)
    public void addEdge(int u, int v, int weight) {
        if (!isWeighted)
            weight = 1;
        matrix[u][v] = weight;
        if (!isDirected) {
            matrix[v][u] = weight;
        }
    }

    // Display adjacency matrix
    public void printMatrix() {
        System.out.println("Adjacency Matrix:");
        for (int i = 0; i < vertices; i++) {
            for (int j = 0; j < vertices; j++) {
                System.out.print(matrix[i][j] + " ");
            }
            System.out.println();
        }
    }

    // Main function to demonstrate usage
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter number of vertices: ");
        int V = scanner.nextInt();

        System.out.print("Is the graph directed? (true/false): ");
        boolean directed = scanner.nextBoolean();

        System.out.print("Is the graph weighted? (true/false): ");
        boolean weighted = scanner.nextBoolean();

        AdjacencyMatrixGraph graph = new AdjacencyMatrixGraph(V, directed, weighted);

        System.out.print("Enter number of edges: ");
        int E = scanner.nextInt();

        System.out.println("Enter edges (format: u v [weight]):");
        for (int i = 0; i < E; i++) {
            int u = scanner.nextInt();
            int v = scanner.nextInt();
            int w = weighted ? scanner.nextInt() : 1;
            graph.addEdge(u, v, w);
        }

        graph.printMatrix();
        scanner.close();
    }
}


/*
Example Input:
Enter number of vertices: 4
Is the graph directed? (true/false): false
Is the graph weighted? (true/false): true
Enter number of edges: 3
Enter edges (format: u v [weight]):
0 1 5
1 2 3
2 3 2

Output:
Adjacency Matrix:
0 5 0 0
5 0 3 0
0 3 0 2
0 0 2 0
 */

File: DSA/GRAPH/1_Basics/Representation/adjacency_matrix_graph.cpp
#include <iostream>
#include <vector>
using namespace std;

class Graph {
private:
    int V;  // Number of vertices
    bool directed;
    bool weighted;
    vector<vector<int>> adjMatrix;

public:
    // Constructor
    Graph(int vertices, bool isDirected = false, bool isWeighted = false) {
        V = vertices;
        directed = isDirected;
        weighted = isWeighted;

        // Initialize matrix with 0 (no edge)
        adjMatrix.resize(V, vector<int>(V, 0));
    }

    // Function to add an edge
    void addEdge(int u, int v, int weight = 1) {
        if (u < 0 || u >= V || v < 0 || v >= V) {
            cout << "Invalid vertex number.\n";
            return;
        }

        if (!weighted) weight = 1;

        adjMatrix[u][v] = weight;

        // If the graph is undirected, also add edge from v to u
        if (!directed) {
            adjMatrix[v][u] = weight;
        }
    }

    // Function to display the adjacency matrix
    void display() {
        cout << "Adjacency Matrix:\n";
        for (int i = 0; i < V; ++i) {
            for (int j = 0; j < V; ++j) {
                cout << adjMatrix[i][j] << " ";
            }
            cout << endl;
        }
    }
};

int main() {
    int vertices = 5;
    bool isDirected = false;
    bool isWeighted = false;

    Graph g(vertices, isDirected, isWeighted);

    // Add some edges
    g.addEdge(0, 1);
    g.addEdge(0, 4);
    g.addEdge(1, 2);
    g.addEdge(1, 3);
    g.addEdge(1, 4);
    g.addEdge(2, 3);
    g.addEdge(3, 4);

    // Display the matrix
    g.display();

    return 0;
}
